### PR TITLE
Pass domain separator as command line argument

### PR DESCRIPTION
**Pull request by e00E**
This has to be configurable for different deployments and us using the
correct value early helps the frontend.

Alex is working on a better version of  that uses the contract address and ethabi's abi encoding. When we have that we can remove the command line argument.

### Test Plan
`cargo run --bin orderbook` prints the domain separator

### CODE REVIEW

**Review by josojo — APPROVED** [review]
Yes, I think it is the right approach to unblock front-end.

My rust is so rusty that it might take a little bit longer. Though, actually it is really not complicated and in js I would have coded it within 2 hours for sure :D 

I will rebase on this Pr...

File: orderbook/src/orderbook.rs
Comment (josojo) [inline orderbook/src/orderbook.rs:81]: nit: you might wanna change the name reflecting that we don't have "user_orders" anymore

**Comment by e00E** [comment]
So if this does depend on the contract address then maybe my approach doesn't make much sense (but we can still keep the DomainSeparator type). We would have to update the value in staging every time we redeploy contracts.
Maybe we could read the value from the contract? But I see it is internal at the moment.

**Comment by fleupold** [comment]
> Maybe we could read the value from the contract? But I see it is internal at the moment.

Let's make it public. I think there are a few private fields which should in fact be publicly accessible for ease of use.

**Comment by nlordell** [comment]
> Let's make it public.

So, while we can make this public, I think we should embrace EIP-712 and compute it in the service. From what I understand from the EIP, this is meant to be computed by the signer or verifyer and has a very well-defined definition.

**Comment by fleupold** [comment]
> I think we should embrace EIP-712 and compute it in the service

In my view the motivation for EIP-712 is to make it more clear to the users what data they are signing. As such any frontend creating a digest to sign (e.g. Metamask), should make sure that when they prompt users with a preview of a structured message, the definition of the format matches with the contract that it is intended for (by manually recomputing the hash or).

Here, we are trying to recover the address from an existing signature. Since this very signature is later to be used with the settlement contract, I'd say its fine to read what domain separator the contract expects rather than computing it ourselves (I don't see an advantage in doing that and it requires more code).

**Comment by nlordell** [comment]
> is to make it more clear to the users what data they are signing

To me the other aspect is intent - the signing Rust code would also explicitly define its expectations on the signing domain and type of the data that was signed and is being recovered.

> I don't see an advantage in doing that and it requires more code

Maybe it is a little overkill, but in my opinion this sort of thing *should* be something that is part of the Rust `web3` library (as it is in the JavaScript `web3` and `ethers` libraries), in which case I would find the code slightly more descriptive to have some call to `eip712::domain_separator(eip712::Domain { ... })` as it would more clearly link the code to the actual signing scheme being used.

The other thing I had thought about before is that computing the domain separator instead of requesting it means one less call that can fail on startup. However, this isn't true as the chain ID needs to be requested anyway.

I guess, I had just envisioned a `eip712` module that would compute all these things and make the calling code very clear on its intent and eventually possibly be added upstream to `web3` crate for others to use, but this is probably quite a bit of overkill.

**Comment by josojo** [comment]
IMO, the domain separator should be public and each project building on gpv2 should decide themselves, whether they want to use it or compute it.

I, personally, would use the domain separator directly from the contract, as it is questionable whether the benefit is worth the additional effort.

**Comment by nlordell** [comment]
> IMO, the domain separator should be public and each project building on gpv2 should decide themselves

Yes I completely agree.

> it is questionable whether the benefit is worth the additional effort

Since the service is only verifying these signatures and not signing, I think you guys are right and there is no real advantage to computing it vs querying it. I think I would feel more strongly about this if there was already a good Rust crate for EIP-712 operations as then I imagine the code to be extremely simple and, at least in my opinion, more clearly describe what the domain separator is exactly and how it should be used (as it would be tied to a technical specification - EIP-712).

**Comment by fleupold** [comment]
> if there was already a good Rust crate for EIP-712

So much ecosystem, so few Fridays...

**Review by bh2smith — APPROVED** [review]
Looks good although I am still not entirely sure what the purpose of the domain separator is? Is this for evaluating anayltics like where the orders are coming from? Is it possible to use a domain separator that would cause the order to be invalid?

**Comment by e00E** [comment]
This PR has been obsoleted as we aret calculating it automatically at start up now.

**Pull request closed without merging**
